lib/avpair.c

```diff
@@ -296,8 +296,6 @@ VALUE_PAIR *rc_avpair_gen(rc_handle const *rh, VALUE_PAIR *pair, unsigned char c
 	if (length != attrlen) {
 		pair = rc_avpair_gen(rh, pair, ptr + attrlen, length - attrlen,
 		    vendorpec);
-		if (pair == NULL)
-			return NULL;
```

Comment:
How is infinite recursion being avoided here? That seems like a very complex recursive code. Would you like to enhance this function a little based on your investigation? Would it make sense to have a more reliable reproducer in the form of `tests/avpair.c`?

Reply:
Yes this code is really hard to follow. I had to single step through it using gdb to figure out how it works.

Infinite recursion is prevented by the test "if (length != attrlen)" at line 296.

It is normal for rc_avpair_gen to return NULL. That's the case, for example, when a Vendor ID or Type cannot be not found in any of the dictionaries. So I don't know why they had the test "if (pair == NULL)". I cannot find a single case where we would want to return when this condition is true.

Reply:
Hi @martin-belanger 

Just a clarification, With this change, will we(radius-client) not process and add the attribute to AV pair eventhough we dont understand it. I assume we can either IGNORE and proceed with processing or STOP and exit the processing. Can we include the VSAs we dont understand ?

Reply:
Hi @aravindprasad90 ,
Per RFC2865, unknown VSAs must be skipped, but the known ones must be processed. Because of the bug (the one I'm fixing here), if the last VSA in the message is unknown (i.e. Vendor ID and/or Type missing from dictionary) then all VSAs in the message get discarded. 

I added a new unit test module. Please run it without my fix. This will demonstrate what I have described above. That is, if a message contains a series of known and unknown VSAs, but the last VSA in the message is for a known Vendor ID and Type then everything is OK. That is, the unknown VSAs get discarded but the known ones are kept. If you repeat the same experiment, but the last VSA in the message in unknown, then all the VSAs (known and unknown) get discarded.

You asked whether we can include VSAs that we don't understand. The code needs the dictionaries to determine data type. Without that the code doesn't know whether it's dealing with a string or integer or something else.